src/sources/statsd/parser.rs

```diff
@@ -29,27 +29,35 @@ pub fn parse(packet: &str) -> Result<Metric, ParseError> {
     let metric_type = parts[1];
 
     let metric = match metric_type {
-        "c" => Metric::Counter {
-            name: sanitize_key(key),
-            val: parts[0].parse()?,
-            sampling: if let Some(s) = parts.get(2) {
-                Some(parse_sampling(s)?)
+        "c" => {
+            let count = if let Some(s) = parts.get(2) {
+                1.0 / sanitize_sampling(parse_sampling(s)?)
             } else {
-                None
-            },
-        },
-        "h" | "ms" => Metric::Timer {
-            name: sanitize_key(key),
-            val: parts[0].parse()?,
-            sampling: if let Some(s) = parts.get(2) {
-                Some(parse_sampling(s)?)
+                1.0
+            };
+            let val: f32 = parts[0].parse()?;
+            let metric = Metric::Counter {
+                name: sanitize_key(key),
+                val: val * count,
+            };
+            metric
+        }
+        "h" | "ms" => {
+            let count = if let Some(s) = parts.get(2) {
+                1.0 / sanitize_sampling(parse_sampling(s)?)
             } else {
-                None
-            },
-        },
-        "g" => Metric::Gauge {
-            name: sanitize_key(key),
-            val: if parts[0]
+                1.0
+            };
+            let val: f32 = parts[0].parse()?;
+            let metric = Metric::Timer {
+                name: sanitize_key(key),
+                val: val * count,
```

Comment:
I'm not sure we want to multiply here. Timers will generally be aggregated into histograms, where we'll want the original timing value. We can just bump the bucket count matching that timing value by `count`. Does that seem right?

Reply:
Timer event in presents of sampling could have two different semantics (like _individual_ vs _cumulative_). The way Statsd represents timer event [is the following](https://github.com/statsd/statsd/blob/master/docs/metric_types.md#timing):

`glork:320|ms|@0.1` => `The glork took 320ms to complete, hey I repeat this message 10 times`

[Cumulative](https://micrometer.io/docs/concepts#_client_side) client-side aggregated approach is to deliver total latency of all requests in the message.

This would mean that we need to convert it to `hey 10 glork invocations took 3200ms in total` message.

The two formats are interchangeable, but I feel like Statsd approach is less popular.

Reply:
That's interesting. I hadn't really thought about client-side aggregation like that. My experience with statsd-style metrics systems is that there really is no client-side aggregation (hence the need for statsd itself). It feels strange to me because it's easy to do the math to disaggregate them, but you'd need some way of knowing that you should.

Since we're in the statsd parser here, it feels like we should stick with the statsd-style (individual sampled measurement instead of cumulative). Does that make sense?

Reply:
You are right, we are parsing Statsd input into `event::Metric` here, which is our internal representation. However I was not sure the latter should be following Statsd semantics, because we will eventually consume other kinds of Timers.

So the question here is what is the meaning of all fields of this structure (from src/event/metrics.rs)
```
Timer {
  name: String,
  val: f32, // total cumulative latency, or latency of each single event?
  count: f32, // count of events in current observation
},
```

Reply:
I think our internal representation should stay focused on individual `event`s and try to deal with aggregations only at the edges. If we later build a source that consumes timers that have been pre-aggregated on the client, it should be simple enough to disaggregate them to match our internal semantics by just dividing `val` by `count`.

Given that, I would say `val` always represents the latency of a single event and `count` is really more like `sample_rate` and represents the number of other events this individual event should be considered representative of. Might even be worth changing to that name to be clearer.

Reply:
Got it. Thanks for making it clear! I've updated the `val` to store individual timing, and renamed `count` to `sample_rate`.

Also looking ahead I've added Histogram type of metric. Histogram is very similar to Timer at the moment, but is a separate type. This way we might have custom processing logic or defaults for timings histograms and arbitrary data histograms. WDYT?

Reply:
That looks good but I'm not sure we need both Timer and Histogram at this point. It might make sense eventually, but I'd rather come up with a concrete use case first. Also, it seems like projects like [SSF](https://github.com/stripe/veneur/tree/master/ssf) have decided to unify the two.

I would probably back out Histogram for now, but then I think we should go ahead and get this merged.

Reply:
Alright. I'm going to close this draftish PR and open a new one #648.